quic/congestion_control/Copa2.h

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ */
+
+#pragma once
+
+#include <quic/congestion_control/third_party/windowed_filter.h>
+#include <quic/state/StateData.h>
+
+namespace quic {
+
+using namespace std::chrono_literals;
+constexpr std::chrono::microseconds kCopa2MinRttWindowLength{10s};
+constexpr std::chrono::microseconds kCopa2ProbeRttInterval{8s};
+
+class Copa2 : public CongestionController {
+ public:
+  explicit Copa2(QuicConnectionStateBase& conn);
+  void onRemoveBytesFromInflight(uint64_t) override;
+  void onPacketSent(const OutstandingPacket& packet) override;
+  void onPacketAckOrLoss(folly::Optional<AckEvent>, folly::Optional<LossEvent>)
+      override;
+
+  [[nodiscard]] uint64_t getWritableBytes() const noexcept override;
+  [[nodiscard]] uint64_t getCongestionWindow() const noexcept override;
+  [[nodiscard]] CongestionControlType type() const noexcept override;
+
+  void setAppIdle(bool, TimePoint) noexcept override;
+  void setAppLimited() override;
+  [[nodiscard]] bool isAppLimited() const noexcept override;
+
+  [[nodiscard]] bool inLossyMode() const noexcept;
+  [[nodiscard]] bool inProbeRtt() const noexcept;
+  [[nodiscard]] uint64_t getBytesInFlight() const noexcept;
+
+ private:
+  void onPacketLoss(const LossEvent&);
+  void onPacketAcked(const AckEvent&);
+  void manageLossyMode(folly::Optional<TimePoint> sentTime);
+
+  QuicConnectionStateBase& conn_;
+  uint64_t cwndBytes_;
+
+  // In packets
+  uint64_t alphaParam_{10};
+  // Loss rate we are willing to tolerate. Actual loss rate will be 2 *
+  // lossTolaranceParam_ + alpha/BDP
+  double lossToleranceParam_{0.05};
+
+  // To get the min RTT over 10 seconds
+  WindowedFilter<
+      std::chrono::microseconds,
+      MinFilter<std::chrono::microseconds>,
+      uint64_t,
+      uint64_t>
+      minRTTFilter_;
+
+  // Updates happen in cycles.
+  folly::Optional<TimePoint> cycleStartTime_{folly::none};
```

Comment:
Optionals default constructed to folly::none. So give it an init none here isn't necessary